src/compiler/checker.ts

```diff
@@ -17336,6 +17336,12 @@ namespace ts {
                         if (right.kind === SyntaxKind.TypeOfExpression && isStringLiteralLike(left)) {
                             return narrowTypeByTypeof(type, <TypeOfExpression>right, operator, left, assumeTrue);
                         }
+                        if (isConstructorAccessExpression(left) && right.kind === SyntaxKind.Identifier) {
+                            return narrowTypeByConstructor(type, left, operator, right, assumeTrue);
+                        }
+                        if (isConstructorAccessExpression(right) && left.kind === SyntaxKind.Identifier) {

```

Comment:
I don't see why in `<expr>.constructor === <expr>` we'd limit what that the syntactic form of `<expr>` was. We're just calling `getTypeOfExpression` anyway.

Reply:
Good point. I'll change this.